blog/2021/08/DEQ.md

```diff
@@ -0,0 +1,426 @@
+a shift in perspective for the machine learning community from discrete to continuous. We
+have always known that powerful techniques such as gradient optimization combined with
+discretized computation gives us the great inference engine that is "deep learning".
+With the addition of NeuralODE, we can take a look at what happens when we take infinitely
+small steps in adjustment, effectively resulting in a differential equation. It is precisely
+this property of NeuralODE that allows it to approximate an "infinitely deep" model
+as each network layer tends to zero in size.
+
+
+As it turns out, there is a line of work following [Deep Equilibrium Models](https://arxiv.org/abs/1909.01377)
+that explores the notion of infinitely-deep neural networks in an explicit manner. It borrows
+much of the ideas from the NeuralODE line of research, and allows neural network structures
+to be viewed as repeated applications of layered computations. This is often very accurate,
+for example when we apply convolution filters on images---the network can essentially
+be viewed as a convolutional feature extractor consisting of repetitive blocks, and one
+linear output layer at the very end. When there is an infinite number of these repeated
+blocks, the network might explode off to infinity. If it doesn't, however, it reaches an
+equilibrium points and converges, hence the name.
+
+## DEQ Models and Infinitely Deep Networks
+
+Let us begin by looking at an instantiation of the equations for DEQ models.
+$$
+\begin{array}{l}
+z:=0\\
+\text { Repeat until convergence: }\\
+z:=\tanh (W z+x)
+\end{array}
+$$
+
+Which eventually brings us to
+$$
+z^{\star}=\tanh \left(W z^{\star}+x\right)
+$$
+
+The idea of DEQ is, in a sense, insanely simple! We assume there are infinitely many
+compositions of identical layers (equivalently an infinitely deep network), and
+directly solve for the fixed point described by this equation. Clearly, naively iterating
+for the fixed point would take an enormous amount of time. Thanks to the Implicit Function
+Theorem, DEQ shows that it is possible to train an infinitely-deep neural network using the
+gradient optimization toolbox. For more reference, please see
+[Deep Implicit Layers](http://implicit-layers-tutorial.org)
+
+From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from
+a differential equations perspective. Thanks to NeuralODE, the machine learning community is now able
+to put on the glasses of differential equations, and see through a new perspective.
+If we perceive the iterating layer of the network as an ODE, DEQ is solving nothing but the
+well-known "Steady State Problem", defined as the final converging state of the ODE as time
+goes to infinity. This line of work is well-studied in the scientific machine learning
+literature, see for example [Notes on Adjoint Methods for 18.335](https://math.mit.edu/~stevenj/18.336/adjoint.pdf)
```

Comment:
```suggestion
directly solve for the fixed point described by this equation. 

Clearly, naively iterating for the fixed point would take an enormous amount of time. However,
classical dynamical systems theory relates this infinite process to a simple mathematical problem.
If we perceive the iterating layer of the network as a dynamical system, DEQ is solving nothing but the
well-known "Steady State Problem", defined as the final converging state of the system as time
goes to infinity. For an update equation $x_{n+1} = x_n + f(x_n)$, "the equation is not changing anymore
is equivalent to saying $x_{n+1} = x_n$ which reduces the problem to finding the $x$ such that $f(x) = 0$.
This is also known as a rootfinding problem, where the forward and adjoint methods have been well-studied
in the scientific computing literature since at least the 90's. For example, Steven Johnson's 
[Notes on Adjoint Methods for 18.335 from 2006](https://math.mit.edu/~stevenj/18.336/adjoint.pdf) shows a
clean derivation of an adjoint equation ("backpropagation" equation) for a rootfinding solver. 
```